Source/Core/VideoBackends/Vulkan/VKTexture.cpp

```diff
@@ -0,0 +1,365 @@
+{
+  _assert_msg_(VIDEO, static_cast<u32>(src_rect.GetWidth()) <= src_texture->GetWidth() &&
+    static_cast<u32>(src_rect.GetHeight()) <= src_texture->GetHeight(),
+    "Source rect is too large for CopyRectangleFromTexture");
+
+  _assert_msg_(VIDEO, static_cast<u32>(dst_rect.GetWidth()) <= m_config.width &&
+    static_cast<u32>(dst_rect.GetHeight()) <= m_config.height,
+    "Dest rect is too large for CopyRectangleFromTexture");
+
+  VkImageCopy image_copy = {
+    { VK_IMAGE_ASPECT_COLOR_BIT, 0, 0,
+    src_texture->GetLayers() },        // VkImageSubresourceLayers    srcSubresource
+    { src_rect.left, src_rect.top, 0 },  // VkOffset3D                  srcOffset
+    { VK_IMAGE_ASPECT_COLOR_BIT, 0, 0,  // VkImageSubresourceLayers    dstSubresource
+    m_config.layers },
+    { dst_rect.left, dst_rect.top, 0 },  // VkOffset3D                  dstOffset
+    { static_cast<uint32_t>(src_rect.GetWidth()), static_cast<uint32_t>(src_rect.GetHeight()),
+    1 }  // VkExtent3D                  extent
+  };
+
+  // Must be called outside of a render pass.
+  StateTracker::GetInstance()->EndRenderPass();
+
+  src_texture->TransitionToLayout(g_command_buffer_mgr->GetCurrentCommandBuffer(),
+    VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
+  m_texture->TransitionToLayout(
+    g_command_buffer_mgr->GetCurrentCommandBuffer(), VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL);
+
+  vkCmdCopyImage(g_command_buffer_mgr->GetCurrentCommandBuffer(), src_texture->GetImage(),
+    VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+    m_texture->GetImage(),
+    VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &image_copy);
+}
+
+void VKTexture::ScaleTextureRectangle(const MathUtil::Rectangle<int>& dst_rect,
+                                      Texture2D* src_texture,
+                                      const MathUtil::Rectangle<int>& src_rect)
+{
+  // Can't do this within a game render pass.
+  StateTracker::GetInstance()->EndRenderPass();
+  StateTracker::GetInstance()->SetPendingRebind();
+
+  // Can't render to a non-rendertarget (no framebuffer).
+  _assert_msg_(VIDEO, m_config.rendertarget,
+    "Destination texture for partial copy is not a rendertarget");
+
+  // Render pass expects dst_texture to be in COLOR_ATTACHMENT_OPTIMAL state.
+  // src_texture should already be in SHADER_READ_ONLY state, but transition in case (XFB).
+  src_texture->TransitionToLayout(g_command_buffer_mgr->GetCurrentCommandBuffer(),
+    VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
+  m_texture->TransitionToLayout(
+    g_command_buffer_mgr->GetCurrentCommandBuffer(), VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
+
+  UtilityShaderDraw draw(g_command_buffer_mgr->GetCurrentCommandBuffer(),
+    g_object_cache->GetPipelineLayout(PIPELINE_LAYOUT_STANDARD),
+    TextureCache::GetInstance()->GetRenderPass(),
+    g_object_cache->GetPassthroughVertexShader(),
+    g_object_cache->GetPassthroughGeometryShader(),
+    TextureCache::GetInstance()->GetEFBColorToTexShader());
```

Comment:
I had to run out quickly this morning, so I only did light testing.  I did see that I picked the wrong shader here.  Thanks 👍